core/drivers/stm32_fmc.c

```diff
@@ -0,0 +1,491 @@
+	fmc_d->conf_data.priv_conf = calloc(1, sizeof(uint32_t));
+	fmc_d->conf_data.lock_conf = calloc(1, sizeof(uint32_t));
+	fmc_d->conf_data.access_mask = calloc(1, sizeof(uint32_t));
+	if (!fmc_d->conf_data.cid_confs || !fmc_d->conf_data.sec_conf ||
+	    !fmc_d->conf_data.priv_conf || !fmc_d->conf_data.access_mask)
+		panic("Missing memory capacity for FMC RIF configuration");
+
+	for (i = 0; i < fmc_d->nb_controller; i++) {
+		rif_conf = fdt32_to_cpu(cuint[i]);
+
+		stm32_rif_parse_cfg(rif_conf, &fmc_d->conf_data,
+				    FMC_NB_MAX_CID_SUPPORTED,
+				    FMC_RIF_CONTROLLERS);
+	}
+
+	fdt_for_each_subnode(ctrl_node, fdt, node) {
+		int status = fdt_get_status(fdt, ctrl_node);
+		uint32_t bank = 0;
+
+		if (status == DT_STATUS_DISABLED)
+			continue;
+
+		if (fdt_read_uint32(fdt, ctrl_node, "reg", &bank) < 0)
+			return TEE_ERROR_BAD_PARAMETERS;
+
+		if (bank != 0)
+			continue;
+
+		if (fdt_getprop(fdt, ctrl_node,
+				"st,fmc2-ebi-cs-cclk-enable", NULL))
+			fmc_d->cclken = true;
+
+		if (!fmc_d->cclken)
+			continue;
+
+		if (fdt_read_uint32(fdt, ctrl_node,
+				    "st,fmc2-ebi-cs-clk-period-ns",
+				    &fmc_d->clk_period_ns) < 0)
+			return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result check_fmc_rif_conf(void)
+{
+	unsigned int i = 0;
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	res = clk_enable(fmc_d->fmc_clock);
+	if (res)
+		panic("Cannot access FMC clock");
+
+	if (fmc_controller_is_secure(0))
+		goto end;
+
+	for (i = 1; i < fmc_d->nb_controller; i++) {
+		uint32_t cidcfgr = io_read32(fmc_d->base + _FMC_CIDCFGR(i));
+		uint32_t semcr = io_read32(fmc_d->base + _FMC_SEMCR(i));
+
+		/* Check if a controller is secure */
+		if (fmc_controller_is_secure(i)) {
+			res = TEE_ERROR_BAD_STATE;
+			goto end;
+		}
+
+		/*
+		 * Check if a controller is shared with incorrect CID
+		 * (!= RIF_CID1)
+		 */
+		res = stm32_rif_check_access(cidcfgr, semcr,
+					     FMC_NB_MAX_CID_SUPPORTED,
+					     RIF_CID1);
+		if (res)
+			break;
+	}
+
+end:
+	clk_disable(fmc_d->fmc_clock);
+
+	return res;
+}
+
+static void configure_fmc(void)
+{
+	uint32_t cidcfgr = 0;
+	uint32_t semcr = 0;
+
+	if (clk_enable(fmc_d->fmc_clock))
+		panic("Cannot access FMC clock");
+
+	semcr = io_read32(fmc_d->base + _FMC_SEMCR(0));
+	cidcfgr = io_read32(fmc_d->base + _FMC_CIDCFGR(0));
+
+	/*
+	 * If OP-TEE doesn't have access to the controller 0,
+	 * then we don't want to try to enable the FMC.
+	 */
+	if (stm32_rif_check_access(cidcfgr, semcr,
+				   FMC_NB_MAX_CID_SUPPORTED, RIF_CID1))
+		goto end;
+
+	/* Check controller 0 access */
+	if (!fmc_controller_is_secure(0)) {
+		DMSG("Controller 0 non-secure, FMC not enabled");
+		goto end;
+	}
+
+	if (cidcfgr & _CIDCFGR_SEMEN &&
+	    stm32_rif_acquire_semaphore(fmc_d->base + _FMC_SEMCR(0),
+					FMC_NB_MAX_CID_SUPPORTED))
+		panic("Couldn't acquire controller 0 semaphore");
+
+	if (fmc_d->pinctrl_d)
+		pinctrl_apply_state(fmc_d->pinctrl_d);
```

Comment:
Sorry, I missed this one: `pinctrl_apply_state()` returns a `TEE_Result` code, it should be checked.
Ditto at line 420.